enhancements/installer/assisted-installer-in-cluster.md

```diff
@@ -0,0 +1,539 @@
+
+* API VIP
+* API VIP DNS name
+* IngressVIP
+
+A new `InstallStrategy` section of the Spec enables the API to describe a way
+of installing a cluster other than the default of using `openshift-install`.
+The new field has an "agent" option that would be common to any Agent-driven
+installation, regardless of platform. Fields include:
+
+* AgentSelector, a label selector for identifying which Agent resources should
+be included in the cluster.
+* ProvisionRequirements, where the API user can specify how many agents to
+expect before beginning installation for each of the control-plane and worker
+roles.
+
+#### Use of metal3
+
+[metal3](https://metal3.io/) will be used to boot the assisted discovery ISO
+for bare-metal deployments. Specifically, the BareMetalHost API will be used
+from a Hub cluster to boot the discovery ISO on hosts that will be used to
+form new clusters.
+
+For this first iteration, it is assumed that the hub cluster itself is using
+the bare metal platform, and thus will have baremetal-operator installed
+and available for use. In the future it may be desirable to make the metal3
+capabilities available on hub clusters that are not using the bare metal
+platform.
+
+#### BareMetalHost can boot live ISOs
+
+A separate [enhancement to
+metal3](https://github.com/metal3-io/metal3-docs/pull/150) proposes a new
+capability in the BareMetalHost API enabling it to boot live ISOs other than
+the Ironic Python Agent. That feature is required so that automation in a
+cluster can boot the discovery ISO on known hardware as the first step toward
+provisioning that hardware.
+
+#### Host Approval
+
+It is important that when an agent checks in, it not be allowed to join a
+cluster until an actor has approved it. From a security standpoint, it is not
+OK for anyone who can access a copy of a discovery ISO and/or access the API
+where agents report themselves to implicitly have the capability to join a
+cluster as a Node.
+
+The Agent CRD will have a field in which to designate that it is approved. If
+the host was booted via the baremetal-operator, approval will be granted
+automatically.
+
+#### Day 2 Add Node Boot-it-Yourself
+
+#### Personas for multi-cluster management
+
+**Infra Owner** Manages physical infrastructure and configures hosts to boot
+the discovery ISO that runs the Agent.
+
+**Cluster Creator** Uses running Agents to create and grow clusters.
+
+#### Create Cluster
+
+This scenario takes place on a hub cluster where hive and possibly RHACM are
+present. This scenario does not include centralized machine management.
+
+1. Infra Owner creates an InstallEnv resource. It can include fields such as egress proxy, NTP server, SSH public key, ... In particular it includes an Agent label selector, and a separate field of labels that should be applied to Agents.
+1. Infra Owner creates BareMetalHost resources that include BMC credentials. They are labeled so that they match the selctor on the InstallEnv.
```

Comment:
Does it matter which namespace the host resources are in?

Using passive voice in "They are labeled..." means it's not clear if the Infra Owner has to make sure to do that, or if something else is going to come along later to do it. I assume the owner has to set the labels correctly themselves? How do they know which to use? Do those match the selectors from step 1, or the separate set of labels in step 1? (Why are both selectors and actual labels needed in the InstallEnv?)

nit: The host resource doesn't include the credentials, those are in a separate Secret resource.

Reply:
> Does it matter which namespace the host resources are in?

Yes. I'll clarify this.
> 
> Using passive voice in "They are labeled..." means it's not clear if the Infra Owner has to make sure to do that, or if something else is going to come along later to do it. I assume the owner has to set the labels correctly themselves?

Yes. I'll clarify this too.

> How do they know which to use? Do those match the selectors from step 1, or the separate set of labels in step 1? 

The selector for step 1 is just for Agents. the InfraEnv and Agent APIs don't know anything about BMHs, because they could be useful to non-bare-metal platforms. So the label on the BMH just needs to be something that the Baremetal Agent Controller (in the next step) recognized as a reference to an InstallEnv. The details of that reference have not been decided yet.

> (Why are both selectors and actual labels needed in the InstallEnv?)

Similar to in a Deployment where you specify both a selector and then have to put a matching label in the PodTemplate. Same for a MachineSet and other similar resources.

> nit: The host resource doesn't include the credentials, those are in a separate Secret resource.

Good point.